opentelemetry-sdk/src/opentelemetry/sdk/metrics/__init__.py

```diff
@@ -0,0 +1,375 @@
+            MeasureHandle(float, self.disabled, self.non_negative))
+        self.handles[label_values] = handle
+        return handle
+
+
+class IntMeasure(metrics_api.IntMeasure):
+    """See `opentelemetry.metrics.IntMeasure`."""
+
+    def __init__(
+        self,
+        name: str,
+        description: str,
+        unit: str,
+        label_keys: List[str] = None,
+        disabled: bool = False,
+        non_negative: bool = False,
+    ) -> None:
+        self.name = name
+        self.description = description
+        self.unit = unit
+        self.label_keys = label_keys
+        self.disabled = disabled
+        self.non_negative = non_negative
+        self.handles = {}
+
+    def get_handle(self,
+                   label_values: List[str]) -> metrics_api.MeasureHandle:
+        """See `opentelemetry.metrics.IntMeasure.get_handle`."""
+        handle = self.handles.get(
+            label_values,
+            MeasureHandle(int, self.disabled, self.non_negative))
+        self.handles[label_values] = handle
+        return handle
+
+
+class CounterHandle(metrics_api.CounterHandle):
+
+    def __init__(
+        self,
+        value_type: Union[Type[float], Type[int]],
+        disabled: bool,
+        non_monotonic: bool) -> None:
+        self.data = 0
+        self.value_type = value_type
+        self.disabled = disabled
+        self.non_monotonic = non_monotonic
+
+    def update(self, value: Union[float, int]) -> None:
+        """See `opentelemetry.metrics.CounterHandle.update`."""
+        self._add(value)
+
+    def _add(self, value: Union[float, int]) -> None:
+        """See `opentelemetry.metrics.CounterHandle._add`."""
+        if self.disabled:
+            logger.warning("Counter metric is disabled.")
+            return
+        if not self.non_monotonic and value < 0:
+            logger.warning("Monotonic counter cannot descend.")
+            return
+        if not isinstance(value, self.value_type):
+            logger.warning("Invalid value passed for %s", self.value_type.__name__)
+            return
+        self.data += value
+
+
+class GaugeHandle(metrics_api.GaugeHandle):
+
+    def __init__(
+        self,
+        value_type: Union[Type[float], Type[int]],
+        disabled: bool,
+        is_monotonic: bool) -> None:
+        self.data = 0
+        self.value_type = value_type
+        self.disabled = disabled
+        self.is_monotonic = is_monotonic
+
+    def update(self, value: Union[float, int]) -> None:
+        """See `opentelemetry.metrics.GaugeHandle.update`."""
+        self._set(value)
+
+    def _set(self, value: Union[float, int]) -> None:
+        """See `opentelemetry.metrics.GaugeHandle._set`."""
+        if self.disabled:
+            logger.warning("Gauge metric is disabled.")
+            return
+        if self.is_monotonic and value < 0:
+            logger.warning("Monotonic gauge cannot descend.")
+            return
+        if not isinstance(value, self.value_type):
```

Comment:
Sorry can you elaborate on this?

Reply:
I mean doing `value = self.value_type(value)` and carry on instead of discarding the value. But I should probably read the spec to see if this is allowed.

Reply:
From what I can see, the spec and RFCs do not explicitly mention what to do in the case of an incorrect value type. It seems to me to be more of an implementation detail. In OpenCensus, a ValueError is raised if the wrong type is passed in. As well, I don't see much point in having different value types if the implementation is simply going to convert the values.

Reply:
The main difference would be how the implementations keep the data internally. Coverting to unlimited precision integer or an double precision floating point number is different.

Reply:
The specs hasn't covered much in terms of implementation details. I believe for this PR, we can proceed as is and once we have more details, we can iterate accordingly.